### ISSUE TITLE
Censored observations in Edward

### ISSUE THREAD

**Issue by kieranrcampbell** [blei-lab/edward#599]
Hi @dustinvtran 

We've been playing around with censored observations in Edward, particularly inference with truncated Weibull distributions, where we know the observed times x are greater than some time o. 

There seems to be an issue in how we let Edward know about the censoring times. For example, if I define a `TruncatedWeibull` class with initialiser

```python
   def __init__(self, shape, # Weibull shape parameter
                    scale, # Weibull scale parameter
                    cens, # Weibull censoring times (o above)
                    ... ):
 ```

followed by 

```python
 parameters = {'shape': shape, 'scale': scale, 'cens': cens}
super(...
            parameters=parameters,
            graph_parents=[self._shape, self._scale, self._cens],
          ...)
```

and construct the model with

```python
x = TruncatedWeibull(..., cens = some_constant_tensor)
```

when we call `inference.run()` the loss exhibits high variance and quickly goes to `NaN`.

However, if I create `TruncatedWeibull2` where I don't let Edward know about the censoring times via an intialiser like

```python
   def __init__(self, shape, # Weibull shape parameter
                    scale, # Weibull scale parameter
                    ... ):
        self._cens = tf.zeros(...)
``` 

then sneakily call

```python
x = TruncatedWeibull(..., cens = some_constant_tensor)
x._cens = some_constant_tensor
```

it appears to work.

I don't know if this is a bug in Edward or my understand is mistaken. I had assumed that I could fix a variable to a constant value (in the same way we can define say x ~ N(0,1) is equivalent to x ~ N(mu, 1) and holding mu=0). Any insight is much appreciated.

A full notebook implementing this is [here](https://nbviewer.jupyter.org/github/kieranrcampbell/edward-weibull/blob/master/test-truncated-weibull.ipynb) and my implementations of TruncatedWeibull and TruncatedWeibull2 are [here](https://github.com/kieranrcampbell/edward-weibull/blob/master/weibull.py).

Thanks,

Kieran

**Comment by dustinvtran** [comment]
Have you verified the behavior across different seeds? My guess is that the inference could just be unstable.

The only difference I can imagine is that when you instantiate `x = TruncatedWeibull(...)`, it will call `_sample_n()` to form the associated tensor for `x`. This means that when you overwrite an attribute, `x._cens = some_constant_tensor`, this would affect future log-prob calls, but it wouldn't affect the way sampling is done during the inference. That said, it looks like you use only `TruncatedWeibull` for the likelihood and you use `ed.KLqp`, so the associated tensor for `x` wouldn't matter.

**Comment by kieranrcampbell** [comment]
I've updated [the notebook](https://github.com/kieranrcampbell/edward-weibull/blob/master/test-truncated-weibull.ipynb) to perform the same thing 100 times for both implementations (for 1000 iterations).

`TruncatedWeibull` gave `NaN`s 46/100 times while `TruncatedWeibull2` only 7/100. Clearly inference is inherently unstable, but that feels like a suspiciously large difference in proportions if the computations being performed are identical. Do you have any thoughts?

**Comment by mariru** [comment]
Hello Kieran,

The problem is that changing the definition of the `_cens` method does not change the sample function. In your notebook you could directly sample from the two Weibull distributions and see that their histograms are different
```
x = TruncatedWeibull(shape = k * tf.ones(N), scale = p * tf.ones(N), cens = censoring_times)

x2 = TruncatedWeibull2(shape = k * tf.ones(N), scale = p * tf.ones(N))
x2._cens = censoring_times

with sess.as_default():
    y = x.sample(100).eval().reshape(-1)
    y2 = x2.sample(100).eval().reshape(-1)
    
plt.hist(y)
plt.hist(y2)
```
To understand what's going on consider this simple tensorflow snippet
```
a = tf.constant(1)
b = tf.constant(5)
c = a + b
b = tf.constant(0)
print(sess.run(c))
```

This will print 6 even though the variable name b has been reassigned to a new tensor. This is because, in the computational graph the node c is still tied to the node that used to be b.

(And in your random variable the sample_n method was defined with the old _cens value. So when you "redefine" _cens a new node is created in the computational graph but all nodes that used to depend on the old _cens, still do.)

**Comment by kieranrcampbell** [comment]
Hi Maja,

Thanks for your help. I see that this would cause different behaviour when sampling from the distributions. In the BBVI example I posted above, I think this sampling is only called for parameter initialisation, and the sampling during inference comes from the approximating distributions? So I still don't see how the differences posted above (and in the updated version) would be explained by this.

I discovered the NaNs in `TruncatedWeibull2` were caused by an x^k term in the log-likelihood, which if `k` is sampled from a log-normal can easily become larger than a float32 can handle. This was solved by constraining `k` to be something reasonable (e.g. between 0 and 5) through a logit bijector.

Thanks again for your help,

Kieran

**Comment by dustinvtran** [comment]
Sure. Thanks for catching your own bug!

**Comment by kieranrcampbell** [comment]
Hi @mariru 

Would you be able to share an example of the histograms you mention? I've implemented

```python
x1 = TruncatedWeibull(shape = 1 * tf.ones(1), scale = 1 * tf.ones(1), cens = 2 * tf.ones(1))
x2 = TruncatedWeibull2(shape = 1 * tf.ones(1), scale = 1 * tf.ones(1))
x2._cens = 2 * tf.ones(1)
sess = tf.Session()

with sess.as_default():
    y1 = x1.sample(100).eval().reshape(-1)
    y2 = x2.sample(100).eval().reshape(-1)
    
plt.hist(y1)
plt.hist(y2)
```

and I get a histogram that looks like 

<img width="375" alt="screen shot 2017-04-11 at 18 33 38" src="https://cloud.githubusercontent.com/assets/2039489/24922212/6e7b9584-1ee5-11e7-99f3-488f206f986c.png">

which look pretty similar, up to sampling noise.

Kieran

**Comment by mariru** [comment]
Kieran,
You are right. I played around again with it and the distributions are the same.
